crates/core_arch/src/wasm32/simd128.rs

```diff
@@ -143,350 +149,738 @@ extern "C" {
+#[inline]
+#[cfg_attr(all(test, all_simd), assert_instr(i64x2.load32x2_u))]
+#[target_feature(enable = "simd128")]
+pub unsafe fn i64x2_load32x2_u(m: *const u32) -> v128 {
+    transmute(simd_cast::<_, u64x2>(*(m as *const u32x2)))
+}
+
+/// Load a single element and splat to all lanes of a v128 vector.
+#[inline]
+#[cfg_attr(all(test, all_simd), assert_instr(v8x16.load_splat))]
+#[target_feature(enable = "simd128")]
+pub unsafe fn v8x16_load_splat(m: *const u8) -> v128 {
+    let v = *m;
+    transmute(u8x16(v, v, v, v, v, v, v, v, v, v, v, v, v, v, v, v))
+}
+
+/// Load a single element and splat to all lanes of a v128 vector.
+#[inline]
+#[cfg_attr(all(test, all_simd), assert_instr(v16x8.load_splat))]
+#[target_feature(enable = "simd128")]
+pub unsafe fn v16x8_load_splat(m: *const u16) -> v128 {
+    let v = *m;
+    transmute(u16x8(v, v, v, v, v, v, v, v))
+}
+
+/// Load a single element and splat to all lanes of a v128 vector.
+#[inline]
+#[cfg_attr(all(test, all_simd), assert_instr(v32x4.load_splat))]
+#[target_feature(enable = "simd128")]
+pub unsafe fn v32x4_load_splat(m: *const u32) -> v128 {
+    let v = *m;
+    transmute(u32x4(v, v, v, v))
+}
+
+/// Load a single element and splat to all lanes of a v128 vector.
+#[inline]
+#[cfg_attr(all(test, all_simd), assert_instr(v64x2.load_splat))]
+#[target_feature(enable = "simd128")]
+pub unsafe fn v64x2_load_splat(m: *const u64) -> v128 {
+    let v = *m;
+    transmute(u64x2(v, v))
 }
 
 /// Stores a `v128` vector to the given heap address.
 #[inline]
 #[cfg_attr(test, assert_instr(v128.store))]
+#[target_feature(enable = "simd128")]
 pub unsafe fn v128_store(m: *mut v128, a: v128) {
-    ptr::write(m, a)
+    *m = a;
 }
 
 /// Materializes a constant SIMD value from the immediate operands.
 ///
 /// The `v128.const` instruction is encoded with 16 immediate bytes
-/// `imm` which provide the bits of the vector directly.
-#[inline]
-#[cfg(not(only_node_compatible_functions))]
-#[rustc_args_required_const(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15)]
-#[cfg_attr(test, assert_instr(
-    v128.const,
-    a0 = 0,
-    a1 = 1,
-    a2 = 2,
-    a3 = 3,
-    a4 = 4,
-    a5 = 5,
-    a6 = 6,
-    a7 = 7,
-    a8 = 8,
-    a9 = 9,
-    a10 = 10,
-    a11 = 11,
-    a12 = 12,
-    a13 = 13,
-    a14 = 14,
-    a15 = 15,
-))]
-pub const fn v128_const(
-    a0: u8,
-    a1: u8,
-    a2: u8,
-    a3: u8,
-    a4: u8,
-    a5: u8,
-    a6: u8,
-    a7: u8,
-    a8: u8,
-    a9: u8,
-    a10: u8,
-    a11: u8,
-    a12: u8,
-    a13: u8,
-    a14: u8,
-    a15: u8,
-) -> v128 {
-    union U {
-        imm: [u8; 16],
-        vec: v128,
-    }
-    unsafe {
-        U {
-            imm: [
-                a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15,
-            ],
-        }
-        .vec
-    }
+/// which provide the bits of the vector directly.
+#[inline]
+#[target_feature(enable = "simd128")]
+pub const unsafe fn v128_const<
```

Comment:
Why isn't this just a normal `const fn` which allows both const and non-const arguments? I see no good reason to restrict ourselves to only allowing constant arguments.

Reply:
Also the naming of `v128_const` feels very inconsistent with the other `*_const` functions. Why not call it `u8x16_const`?

Reply:
Ah this is a restriction of wasm where the `v128.const` instruction does not take any runtime arguments, so all of the arguments to this function are required to be constant. If anything is non-const then instruction-wise you'd have to use the `*_replace_lane` functions.

Reply:
Sure but we can offer more flexibility in the intrinsics and leave the choice of the exact instruction sequence to use to the compiler. As a general rule we do not make any guarantees about the exact instruction sequence that an intrinsic will generate (use inline asm if you need that).

FWIW all the other architectures (x86 SSE, ARM NEON) provide a function to construct a vector from elements, and I think we should do the same here.

Reply:
Seems reasonable to me. I agree it's much nicer to *not* have to use const generics, and I think it's pretty reasonable to say that compilers are either smart enough to use `v128.const` or they'll do the manual replace/extract lane business.